### ISSUE TITLE
Frame throttling suffers from blocking if time shifts backwards.

### ISSUE THREAD

**Issue by lawrence-dol** [lhorie/mithril.js#581]
In practice, this might be internally compensated due to forced redrawing, but...

In theory, if time is adjusted backwards between throttled frames (e.g. NTP resets the clock), redraws will stop until the clock catches up.

The offending line is:

    if (new Date - lastRedrawCallTime > FRAME_BUDGET || $requestAnimationFrame === window.requestAnimationFrame) {

(which should be `Date.now()` not `new Date`, BTW -- why construct an object to extract the UTC datestamp).

Backward movement could be detected and the lastRedrawCallTime reset to compensate:

    function currentTime() {
        var tim=Date.now();
        if(tim<lastKnownTime) {
            lastRedrawCallTime=tim;
            // and any others
        }
        lastKnownTime=tim;
        return tim;
    }

Alternatively, instead of using Date it could use `performance.now()` which is a guaranteed advancing clock that seems to be well supported in evergreen browsers.

https://developer.mozilla.org/en-US/Add-ons/Code_snippets/Timers

    Feature         Chrome          Firefox (Gecko)     Internet Explorer   Opera       Safari
    Basic support   20.0 webkit     15.0 (15.0)         10.0                15.0        8.0
                    24.0 [1]

**Comment by markmarijnissen** [comment]
The alternative is to use the `performance` api to get seconds-on-page.

**Comment by barneycarroll** [comment]
The code in question is only ever invoked if requestAnimationFrame isn't available — and all the APIs that mitigate the issue supersede requestAnimationFrame :/

**Comment by lawrence-dol** [comment]
Closing as: Won't / can't fix.

**Comment by isiahmeadows** [comment]
I'd like to add to that that `new Date - someNumber` is equivalent to `new Date().valueOf() - someNumber`, which is also equivalent to `Date.now() - someNumber`, since `Date.prototype.valueOf` returns the value of the date in milliseconds, which was constructed to be the current time value. There's little difference, and is really just taking advantage of the standard implicit numeric conversion in JS and the API. I float between `+new Date()` and `Date.now()` in practice, since the two are completely equivalent.

**Comment by lawrence-dol** [comment]
Because `Date.now()` "is a static method of Date" it's theoretically more efficient because a Date object doesn't need to be constructed.

**Comment by isiahmeadows** [comment]
In practice, the performance is identical, even in microbenchmarks, with or without the method cached.

**Comment by lawrence-dol** [comment]
But with a possibly higher cost in CPU, memory, GC load, etc. Which will tend to matter more the less powerful the hardware.

**Comment by isiahmeadows** [comment]
@lawrence-dol

TL;DR: It doesn't really matter.

---

The tick delay isn't performance sensitive, anyways (the redraw is done synchronously, and takes a *lot* more time). Plus, memory leaks are trivial to avoid.

Although now I'm actually running it in Node, the `Date.now` approach is usually about 3 *microseconds* faster (~6 μs vs ~3 μs), but a single redraw even on small apps usually takes about 50 *milliseconds* minimum. The TodoMVC benchmark on my machine usually takes about 8.5 sec for adding+checking+removing 500 items after about 5-10 runs for conditioning. That equates to about 1500 redraws in 8.5 sec, or about 5.7 ms for each redraw, using `requestAnimationFrame`. Since the fallback uses `setTimeout`, that's going to be even slower than the few microsecond difference. If you care about something that small, you should *still* prefer nanosecond timers like `performance.now` (or `process.hrtime` in Node) if possible.

Or in other words, *it doesn't matter*.

**Comment by isiahmeadows** [comment]
@lawrence-dol Mind if we continue this discussion [privately in Gitter](https://gitter.im/isiahmeadows)?
